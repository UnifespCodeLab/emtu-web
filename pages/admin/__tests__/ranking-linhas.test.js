import { mount, createLocalVue, RouterLinkStub } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Vuex, { Store } from 'vuex'
import emtuApi from '~/assets/services/emtu-api'

import AdminLineRanking from '~/pages/admin/ranking-linhas.vue'

jest.requireActual('~/assets/services/emtu-api')

const localVue = createLocalVue()
localVue.use(Vuex)

let vuetify
let wrapper
class ResizeObserver {
  observe () {}
  unobserve () {}
}

const endDate = new Date(Date.now() - new Date().getTimezoneOffset() * 60000)
  .toISOString()
  .substr(0, 10)

const store = new Store({
  modules: {
    cid: {
      namespaced: true,
      state: {
        cids: []
      },
      actions: {
        fetchCids: () => []
      }
    },
    alert: {
      namespaced: true,
      actions: {
        showAlert: jest.fn()
      }
    }
  }
})

describe('Pages / Admin / AdminLineRanking', () => {
  beforeEach(() => {
    jest.mock('vue-apexcharts', () => jest.fn())
    window.ResizeObserver = ResizeObserver

    vuetify = new Vuetify()

    wrapper = mount(AdminLineRanking, {
      localVue,
      vuetify,
      store,
      stubs: {
        'router-link': RouterLinkStub
      }
    })
  })

  it('should render the page', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render the correct elements', () => {
    const selects = wrapper.findAllComponents({ name: 'v-select' })
    expect(selects.length).toBe(2)

    const dialogs = wrapper.findAllComponents({ name: 'v-dialog' })
    expect(dialogs.length).toBe(2)

    const textFields = wrapper.findAllComponents({ name: 'v-text-field' })
    expect(textFields.length).toBe(3)

    const datePickers = wrapper.findAllComponents({ name: 'v-date-picker' })
    expect(datePickers.length).toBe(0)

    const searchButton = wrapper.findComponent({ name: 'v-btn' })
    expect(searchButton.props('color')).toBe('primary')
  })

  it('should render the correct start date', () => {
    const textField = wrapper.findAllComponents({ name: 'v-text-field' }).at(0)
    expect(textField.props('value')).toBe(null)
  })

  it('should render the correct end date', () => {
    const textField = wrapper.findAllComponents({ name: 'v-text-field' }).at(2)
    expect(textField.props('value')).toBe(endDate)
  })

  it('should not render the chart container', () => {
    const container = wrapper.find('.line-ranking__graph-wrapper')
    expect(container.exists()).toBe(false)
  })

  describe('when search button is clicked', () => {
    beforeEach(() => {
      emtuApi.get = jest.fn().mockResolvedValue({
        data: {
          data: [{
            idLinha: '244',
            searchCount: 10
          },
          {
            idLinha: '420',
            searchCount: 5
          }]
        }
      })

      const searchButton = wrapper.findAllComponents({ name: 'v-btn' }).at(0)
      searchButton.trigger('click')
    })

    it('calls emtu-api to execute ranking search', () => {
      expect(emtuApi.get).toHaveBeenCalledWith('/searches/ranking', {
        params: {
          sucedida: null,
          startDate: null,
          endDate,
          idCid: null,
          limite: null
        }
      })
    })

    it('should render the chart container', async () => {
      await wrapper.vm.$nextTick()

      const container = wrapper.find('.line-ranking__graph-wrapper')
      expect(container.exists()).toBe(true)
    })

    it('should render correct chart', async () => {
      await wrapper.vm.$nextTick()

      const chart = wrapper.findComponent({ name: 'AdminChart' })
      expect(chart.props('title')).toBe('Ranking de linhas')
      expect(chart.props('type')).toBe('bar')
      expect(chart.props('categories')).toEqual(['244', '420'])
    })
  })
})
