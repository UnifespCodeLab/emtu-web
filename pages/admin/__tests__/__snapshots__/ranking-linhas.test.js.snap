// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages / Admin / AdminLineRanking should render the page 1`] = `
<div
  class="line-ranking"
>
  <div
    class="line-ranking__search-container"
  >
    <div
      class="v-input theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--enclosed v-select"
    >
      <div
        class="v-input__control"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="list-2"
          class="v-input__slot"
          role="button"
        >
          <div
            class="v-select__slot"
          >
            <label
              class="v-label theme--light"
              for="input-2"
              style="left: 0px; position: absolute;"
            >
              Resultado Buscas
            </label>
            <div
              class="v-select__selections"
            >
              <input
                aria-readonly="false"
                autocomplete="off"
                id="input-2"
                readonly="readonly"
                type="text"
              />
            </div>
            <div
              class="v-input__append-inner"
            >
              <div
                class="v-input__icon v-input__icon--append"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-menu-down theme--light"
                />
              </div>
            </div>
            <input
              type="hidden"
            />
          </div>
          <div
            class="v-menu"
          >
            <!---->
          </div>
        </div>
        <div
          class="v-text-field__details"
        >
          <div
            class="v-messages theme--light"
          >
            <transition-group-stub
              class="v-messages__wrapper"
              name="message-transition"
              tag="div"
            />
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="v-input theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--enclosed"
    >
      <div
        class="v-input__control"
      >
        <div
          class="v-input__slot"
        >
          <div
            class="v-text-field__slot"
          >
            <label
              class="v-label theme--light"
              for="input-7"
              style="left: 0px; position: absolute;"
            >
              Limite
            </label>
            <input
              id="input-7"
              max="10"
              type="number"
            />
          </div>
        </div>
        <div
          class="v-text-field__details"
        >
          <div
            class="v-messages theme--light"
          >
            <transition-group-stub
              class="v-messages__wrapper"
              name="message-transition"
              tag="div"
            />
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="v-input theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--enclosed v-select"
    >
      <div
        class="v-input__control"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="list-10"
          class="v-input__slot"
          role="button"
        >
          <div
            class="v-select__slot"
          >
            <label
              class="v-label theme--light"
              for="input-10"
              style="left: 0px; position: absolute;"
            >
              Cid
            </label>
            <div
              class="v-select__selections"
            >
              <input
                aria-readonly="false"
                autocomplete="off"
                id="input-10"
                readonly="readonly"
                type="text"
              />
            </div>
            <div
              class="v-input__append-inner"
            >
              <div
                class="v-input__icon v-input__icon--append"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-menu-down theme--light"
                />
              </div>
            </div>
            <input
              type="hidden"
            />
          </div>
          <div
            class="v-menu"
          >
            <!---->
          </div>
        </div>
        <div
          class="v-text-field__details"
        >
          <div
            class="v-messages theme--light"
          >
            <transition-group-stub
              class="v-messages__wrapper"
              name="message-transition"
              tag="div"
            />
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="v-dialog__container"
    >
      <!---->
    </div>
    <div
      class="v-input daily-acces__input theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--enclosed v-text-field--placeholder"
    >
      <div
        class="v-input__control"
      >
        <div
          class="v-input__slot"
        >
          <div
            class="v-text-field__slot"
          >
            <label
              class="v-label theme--light"
              for="input-16"
              style="left: 0px; position: absolute;"
            >
              Data in√≠cio
            </label>
            <input
              aria-expanded="false"
              aria-haspopup="true"
              id="input-16"
              role="button"
              type="text"
            />
          </div>
          <div
            class="v-input__append-inner"
          >
            <div />
          </div>
        </div>
        <div
          class="v-text-field__details"
        >
          <div
            class="v-messages theme--light"
          >
            <transition-group-stub
              class="v-messages__wrapper"
              name="message-transition"
              tag="div"
            />
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="v-dialog__container"
    >
      <!---->
    </div>
    <div
      class="v-input daily-acces__input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--enclosed v-text-field--placeholder"
    >
      <div
        class="v-input__control"
      >
        <div
          class="v-input__slot"
        >
          <div
            class="v-text-field__slot"
          >
            <input
              aria-expanded="false"
              aria-haspopup="true"
              id="input-20"
              readonly="readonly"
              role="button"
              type="text"
            />
          </div>
          <div
            class="v-input__append-inner"
          >
            <div
              class="v-input__icon v-input__icon--clear"
            >
              <button
                aria-label="clear icon"
                class="v-icon notranslate v-icon--link mdi mdi-close theme--light"
                type="button"
              />
            </div>
          </div>
        </div>
        <div
          class="v-text-field__details"
        >
          <div
            class="v-messages theme--light"
          >
            <transition-group-stub
              class="v-messages__wrapper"
              name="message-transition"
              tag="div"
            />
          </div>
        </div>
      </div>
    </div>
     
    <button
      class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--large primary"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        
      BUSCAR
    
      </span>
    </button>
  </div>
   
  <!---->
</div>
`;
